File: src/hollywood/archive.py
from __future__ import annotations

import hashlib
import re
from datetime import UTC

from .config import HollywoodSettings
from .models import ArchivedPayload, RawPayload, SourceDefinition, json_dumps, make_stable_id

SAFE_NAME_RE = re.compile(r"[^a-zA-Z0-9._-]+")


def _safe_name(value: str) -> str:
    cleaned = SAFE_NAME_RE.sub("-", value.strip())
    return cleaned.strip("-") or "payload"


def _extension(payload: RawPayload) -> str:
    if payload.extension:
        return payload.extension
    if payload.payload_type.endswith("xml"):
        return ".xml"
    if payload.payload_type.endswith("html"):
        return ".html"
    if payload.payload_type.endswith("text"):
        return ".txt"
    if payload.payload_type.endswith("json"):
        return ".json"
    if payload.payload_type.endswith("tsv"):
        return ".tsv"
    return ".bin"


def archive_payload(
    settings: HollywoodSettings,
    source: SourceDefinition,
    payload: RawPayload,
) -> ArchivedPayload:
    settings.ensure_directories()
    fetched = payload.fetched_at.astimezone(UTC)
    day_dir = settings.raw_dir / source.source_id / fetched.strftime("%Y/%m/%d")
    day_dir.mkdir(parents=True, exist_ok=True)

    content_hash = hashlib.sha256(payload.body).hexdigest()
    filename = (
        f"{fetched.strftime('%H%M%S')}_{_safe_name(payload.logical_id)[:80]}{_extension(payload)}"
    )
    path = day_dir / filename
    if not path.exists():
        path.write_bytes(payload.body)

    raw_record_id = make_stable_id(
        source.source_id, payload.payload_type, payload.logical_id, content_hash
    )
    return ArchivedPayload(
        raw_record_id=raw_record_id,
        source_id=source.source_id,
        source_kind=source.kind.value,
        payload_type=payload.payload_type,
        logical_id=payload.logical_id,
        content_path=str(path),
        content_hash=content_hash,
        content_type=payload.content_type,
        source_url=payload.source_url,
        canonical_url=payload.canonical_url,
        fetched_at=payload.fetched_at,
        metadata_json=json_dumps(payload.metadata),
    )

File: api/src/ingest/archive.ts
import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { resolve } from 'node:path';

import { env } from '../env.js';
import { makeStableId } from './models.js';
import type { ArchivedPayload, RawPayload, SourceDefinition } from './models.js';

const SAFE_NAME_RE = /[^a-zA-Z0-9._-]+/g;

function safeName(value: string): string {
  const cleaned = value.trim().replace(SAFE_NAME_RE, '-');
  return cleaned.replace(/^-+|-+$/g, '') || 'payload';
}

function extension(payload: RawPayload): string {
  if (payload.extension) return payload.extension;
  if (payload.payloadType.endsWith('xml')) return '.xml';
  if (payload.payloadType.endsWith('html')) return '.html';
  if (payload.payloadType.endsWith('text')) return '.txt';
  if (payload.payloadType.endsWith('json')) return '.json';
  if (payload.payloadType.endsWith('tsv')) return '.tsv';
  return '.bin';
}

function pad(n: number): string {
  return String(n).padStart(2, '0');
}

export function archivePayload(source: SourceDefinition, payload: RawPayload): ArchivedPayload {
  const fetched = payload.fetchedAt;
  const dayDir = resolve(
    env.HOLLYWOOD_DATA_DIR,
    'raw',
    source.sourceId,
    String(fetched.getUTCFullYear()),
    pad(fetched.getUTCMonth() + 1),
    pad(fetched.getUTCDate()),
  );
  mkdirSync(dayDir, { recursive: true });

  const contentHash = createHash('sha256').update(payload.body).digest('hex');
  const timePart = `${pad(fetched.getUTCHours())}${pad(fetched.getUTCMinutes())}${pad(fetched.getUTCSeconds())}`;
  const filename = `${timePart}_${safeName(payload.logicalId).slice(0, 80)}${extension(payload)}`;
  const path = resolve(dayDir, filename);
  if (!existsSync(path)) {
    writeFileSync(path, payload.body);
  }

  const rawRecordId = makeStableId(
    source.sourceId,
    payload.payloadType,
    payload.logicalId,
    contentHash,
  );
  return {
    rawRecordId,
    sourceId: source.sourceId,
    sourceKind: source.kind,
    payloadType: payload.payloadType,
    logicalId: payload.logicalId,
    contentPath: path,
    contentHash,
    contentType: payload.contentType,
    sourceUrl: payload.sourceUrl,
    canonicalUrl: payload.canonicalUrl,
    fetchedAt: payload.fetchedAt,
    metadataJson: JSON.stringify(payload.metadata),
  };
}
